cob_obstacle_distance/src/distance_manager.cpp

```diff
@@ -0,0 +1,386 @@
+            fcl::DistanceResult tmpResult;
+            fcl::DistanceRequest distRequest(true, 1.0, 100.0);
+            fcl::FCL_REAL dist = fcl::distance(&ooi_co, &collision_obj, distRequest, tmpResult);
+            if (dist < lastDist)
+            {
+                setDistResult = true;
+                dist_result = tmpResult;
+                result_collision_obj = collision_obj;
+            }
+        }
+
+        if(setDistResult)
+        {
+            ROS_INFO_STREAM("Minimum distance: " << dist_result.min_distance);
+            fcl::Vec3f t = dist_result.nearest_points[1] + result_collision_obj.getTranslation(); // Translation from "base_link" frame!!!
+            fcl::Quaternion3f q = result_collision_obj.getQuatRotation();
+
+            Eigen::Vector3d obst_vector(t[VEC_X],
+                                       t[VEC_Y],
+                                       t[VEC_Z]);
+
+            ROS_INFO_STREAM("absJntPosition: " << abs_jnt_pos);
+            Eigen::Vector3d dist_vector = abs_jnt_pos - obst_vector;
+            cob_obstacle_distance::ObstacleDistance od_msg;
+            od_msg.distance = dist_result.min_distance;
+            od_msg.distance_vector.x = dist_vector(VEC_X);
+            od_msg.distance_vector.y = dist_vector(VEC_Y);
+            od_msg.distance_vector.z = dist_vector(VEC_Z);
+            od_msg.id.frame_id = frame_of_interest_name;
+            od_msg.id.stamp = ros::Time::now();
+            obstacle_distances.distances.push_back(od_msg);
+        }
+    }
+
+    if(obstacle_distances.distances.size() > 0)
+    {
+        this->obstacle_distances_pub_.publish(obstacle_distances);
+    }
+}
+
+bool DistanceManager::transform()
+{
+    bool success = true;
+    try
+    {
+        tf::StampedTransform cb_transform_bl;
+        tf_listener_.waitForTransform(chain_base_link_, "base_link", ros::Time(0), ros::Duration(0.5));
+        tf_listener_.lookupTransform(chain_base_link_, "base_link", ros::Time(0), cb_transform_bl);
+        tf::transformTFToEigen(cb_transform_bl, tf_cb_frame_bl_);
+    }
+    catch (tf::TransformException &ex)
+    {
+        success = false;
+        ROS_ERROR("%s",ex.what());
+    }
+
+    return success;
+}
+
+void DistanceManager::jointstateCb(const sensor_msgs::JointState::ConstPtr& msg)
+{
+    KDL::JntArray q_temp = last_q_;
+    KDL::JntArray q_dot_temp = last_q_dot_;
+    uint16_t count = 0;
+
+    for(uint16_t j = 0; j < chain_.getNrOfJoints(); ++j)
+    {
+        for(uint16_t i = 0; i < msg->name.size(); i++)
+        {
+            if(strcmp(msg->name[i].c_str(), joints_[j].c_str()) == 0)
+            {
+                q_temp(j) = msg->position[i];
+                q_dot_temp(j) = msg->velocity[i];
+                count++;
+                break;
+            }
+        }
+    }
+
+    if(joints_.size() == count)
+    {
+        last_q_ = q_temp;
+        last_q_dot_ = q_dot_temp;
+    }
+    else
+    {
+        ROS_ERROR("Failed to jointstate_cb");
+    }
```

Comment:
every chain also publishes its joint states on a namespaced topic, i.e. there is also a /arm_right/joint_states that only consists of the joint_state of the joints of arm_right. If you subscribe to ```joint_states``` instead of ```/joint_states``` (using appropriate nh), no need for parsing!